### ISSUE TITLE
The reproducibility of this paper

### ISSUE THREAD

**Issue by Hanminghao** [bowang-lab/BLEEP#12]
Hello author, first of all thank you for your outstanding work. I had a problem with poor test set results while experimenting entirely with your code and data. Specifically, the model is running on a single A6000 48GB GPU. I have set the batch size to 512 and the learning rate to 1e-3. Apart from not using distributed training, my setup is the same as yours. In addition, the best model filtered by the validation set appears at epoch 38, with a val loss of 3.38. However, the model is only showing performance around 0.02 and 0.03 on the metrics of mean correlation highly expressed genes and mean correlation highly variable genes. I really look forward to your receiving my questions. Thank you.
![image](https://github.com/user-attachments/assets/f3e691ea-1b31-4783-b9a2-ed103656636d)
![image](https://github.com/user-attachments/assets/f8596b1e-ecd1-41ca-9a89-b81173245175)

**Comment by Hanminghao** [comment]
After I changed to distributed training, I got results similar to the manuscript.
![image](https://github.com/user-attachments/assets/78785670-1645-4e09-abac-1aab17d69fc9)

**Comment by NBitBuilder** [comment]
Hi, @Hanminghao ,

I encountered the same issue with you about the low correlation values while training on an A6000. Have you been able to identify the cause? Any insights would be appreciated

**Comment by NBitBuilder** [comment]
Why does distributed training have such a significant impact? I also noticed that the contrastive loss differs from methods like MoCo-v3, as it doesn't gather samples from other GPUs for loss computation. Could you also share the batch size you used for each GPU?

**Comment by Hanminghao** [comment]
I'm sorry to hear that, but in my many tests, I can get good results without using distributed training. My specific operation was to set the batch size to 128, and the following results could be obtained when training on a single A6000: HVG 10.27 HEG:18.97. Note that I did not test only on a single slide as set in the original text, but tested on four slides respectively and calculated the average value.

**Comment by NBitBuilder** [comment]
Thank you for your response. I’m now getting reasonable results using a batch size of 128 and gradient accumulation of 4. This setup mimics bsz=512 on distributed training by reducing the loss instead of concatenating logits before calculating the loss.

I’m quite surprised by how sensitive this type of model is to hyperparameters like batch size.

**Comment by NBitBuilder** [comment]
| Model                | Mean Correlation (Cells) | Max Correlation | Mean HEG | Mean HVG | Mean Markers |
|----------------------|--------------------------|-----------------|----------|----------|--------------|
| BLEEP (bsz=128, accum=4) | 0.8025                   | 0.6810          | 0.1630   | 0.1657   | 0.2280       |
| BLEEP (bsz=128, accum=1) | 0.7149                   | 0.6282          | 0.1096   | 0.0988   | 0.1158       |

**Comment by NBitBuilder** [comment]
Additionally, the image transformation for ResNet50 in the repository code is incorrect. Proper normalization needs to be applied, which significantly improves performance and results in faster convergence, reducing the required training time to less than 30 epochs.


```python
    def transform(self, image):
        image = Image.fromarray(image)
        
        # Random flipping and rotations
        if random.random() > 0.5:
            image = TF.hflip(image)
        if random.random() > 0.5:
            image = TF.vflip(image)
        angle = random.choice([180, 90, 0, -90])
        image = TF.rotate(image, angle)
        
        # Convert to tensor
        image = TF.to_tensor(image)
        
        # Normalize using ImageNet mean and std
        image = TF.normalize(image, mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225])
        
        return image
```

| Metric                  | BLEEP (bsz=128, accum=4) | BLEEP (bsz=128, accum=1) | BLEEP (bsz=128, accum=4) normalized |
|-------------------------|--------------------------|--------------------------|--------------------------------------|
| Mean correlation cells   | 0.8025                   | 0.7149                   | 0.8051                               |
| Max correlation          | 0.6810                   | 0.6282                   | 0.6913                               |
| Mean HEG                 | 0.1630                   | 0.1096                   | 0.1977                               |
| Mean HVG                 | 0.1657                   | 0.0988                   | 0.1862                               |
| Mean markers             | 0.2280                   | 0.1158                   | 0.2397                               |
